### PR TITLE
Added multi datastore support.

### LINKED ISSUE #288: Spec volume placement on different datastores
We need to design volume lifecycle which allows placement on different datastores. 
One use case brought up by @Govint is "I want to have all volumes in one place". Another is "VM is relocated but still want access to old volumes". 
We should design and discuss the 'datastore usage' - how volumes are created on different datastores, listed, used, deleted - when using client or server `-o datastore=<name>` options; We then should implement it for GA

### PR DESCRIPTION
**Pull request by msterin**
#Fixes #288 

### Design
- each volume has short name (same as today) and fully qualified name (volume@datastore).
- the '@datastore' part can be omitted , and it means "use current VM datastore'
- default behavior (when using short names) is exactly as today - no changes
- on 'create' , a volume can be created as '--name=volume@datastore' . The result is a creation of 'volume.vmdk' in `datastore`
  -  ' -o datastore=datastore' is not supported yet
  - if datastore is omitted, 'volume' is created on VMs's datastore and 'volume' is returned to Docker
  - if datastore is present, 'volume' is created on this store and 'volume@datastore' is returned to Docker
  - if datastore does not exits, a list of existing ones is returned in error message
- 'list' operation shows short names for volumes on current datastore, and fully qualified names for others. E.g
   $ docker volume ls
   myVol                vmdk
   myVol@vsanDatastore  vmdk
   someOtherVolume      vmdk

### Test
* basic test automation for name parse is included
* test automation for create/delete based on parsed name is not included - needs agreement on datastore names, and general rethinking/redesign or automation suit and libs (see #472 )
It would be too much to do now, so tested it manually (below)
* `make all` and CI

Manual test: 
* create/delete volumes with default datastore (vsanDatastore) and not default one (datastore266) but the same name
* do create/delete on volume and make sure the right one is impacted when default datastotre is used, and when non-default datastore is used. 
*  run container ans make sure the volumes are different.

Test log:
```
root@photon-machine [ ~ ]# docker volume ls 
DRIVER              VOLUME NAME
vmdk                myVol33@datastore226
vmdk                superduper
vmdk                superduper@datastore226
vmdk                vol
vmdk                vol1
root@photon-machine [ ~ ]# docker volume create -d vmdk --name=volume@badstore
Error response from daemon: create volume@badstore: VolumeDriver.Create: Invalid datastore 'badstore'.
Known datastores: datastore226, build-toolchain, exit14.eng.vmware.com, build-apps, pa-dbc1122.eng.vmware.com, scm-trees, vsanDatastore.
Default datastore: vsanDatastore
root@photon-machine [ ~ ]# docker volume create --driver=vmdk --name=vol@vsanDatastore
Error response from daemon: create vol@vsanDatastore: VolumeDriver.Create: File /vmfs/volumes/vsanDatastore/dockvols/vol.vmdk already exists
root@photon-machine [ ~ ]# docker volume create --driver=vmdk --name=vol@datastore226 
vol@datastore226
root@photon-machine [ ~ ]# docker volume ls 
DRIVER              VOLUME NAME
vmdk                myVol33@datastore226
vmdk                superduper
vmdk                superduper@datastore226
vmdk                vol
vmdk                vol1
vmdk                vol@datastore226
root@photon-machine [ ~ ]# docker volume rm superduper
superduper
root@photon-machine [ ~ ]# docker volume ls 
DRIVER              VOLUME NAME
vmdk                myVol33@datastore226
vmdk                superduper@datastore226
vmdk                vol
vmdk                vol1
vmdk                vol@datastore226
root@photon-machine [ ~ ]# docker volume rm superduper@datastore226
superduper@datastore226
root@photon-machine [ ~ ]# docker volume ls 
DRIVER              VOLUME NAME
vmdk                myVol33@datastore226
vmdk                vol
vmdk                vol1
vmdk                vol@datastore226
root@photon-machine [ ~ ]# docker run --rm -v vol:/v busybox touch /v/MyFile
root@photon-machine [ ~ ]# docker run --rm -v vol:/v busybox ls /v
MyFile
lost+found
root@photon-machine [ ~ ]# docker run --rm -v vol@datastore226:/v busybox ls /v          
lost+found
root@photon-machine [ ~ ]# docker run --rm -v vol@datastore226:/v busybox touch /v/HisFile
root@photon-machine [ ~ ]# docker run --rm -v vol@datastore226:/v busybox ls /v
HisFile
lost+found
root@photon-machine [ ~ ]# docker run --rm -v vol:/v busybox ls /v
MyFile
lost+found

```

### CODE REVIEW

**Comment by andrewjstone** [comment]
@msterin Did you do any manual tests with the Admin CLI to ensure it all still works as expected?

**Comment by msterin** [comment]
@andrewjstone not before you asked. There is nothing really impacted in admin cli, but I did run it now and it all good - at least `ls` command , See below. What else do you think is worth checking ? 
Docker:
```
root@photon-machine [ ~ ]# docker volume ls 
DRIVER              VOLUME NAME
vmdk                myVol33@datastore226
vmdk                vol
vmdk                vol1
vmdk                vol@datastore226
```
Admin CLI:
```
root@promc-2n-dhcp35> /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py  ls
Volume   Datastore      Created By VM  Created                   Attached To VM  Policy          Capacity  Used     
-------  -------------  -------------  ------------------------  --------------  --------------  --------  -------  
myVol33  datastore226   PhotonRC_1     Wed Jun 15 01:12:29 2016  detached        N/A             100.00MB  14.00MB  
vol      datastore226   PhotonRC_1     Wed Jun 15 21:22:13 2016  detached        N/A             100.00MB  14.00MB  
vol      vsanDatastore  Photon1RC      Tue Jun 14 01:23:06 2016  PhotonRC_1      [VSAN default]  1.00GB    40.00MB  
vol1     vsanDatastore  PhotonRC_1     Tue Jun 14 18:02:02 2016  detached        [VSAN default]  100.00MB  56.00MB  
```

**Comment by andrewjstone** [comment]
@msterin That seems good enough :)

File: esx_service/utils/vmdk_utils.py [hunk 1]
```diff
@@ -50,19 +55,23 @@ def get_datastores():
...
+    """ forms full name as <path-to-volumes>/<volname>.vmdk"""
+    return os.path.join(path, "%s.vmdk" % vol_name)
+
+
 def list_vmdks(path):
     """ Return a list all VMDKs in a given path """
     try:
-        files = os.listdir(path)
-        return [f for f in files
+        return [f for f in os.listdir(path)
```
Comment (andrewjstone) [inline esx_service/utils/vmdk_utils.py]: I like this change :)

File: esx_service/utils/vmdk_utils.py [hunk 2]
```diff
@@ -50,19 +55,23 @@ def get_datastores():
...
         for file_name in list_vmdks(path):
             volumes.append({'path': path,
                             'filename': file_name,
                             'datastore': datastore})
     return volumes
 
 
+def get_vmdk_name(path, vol_name):
+    """ forms full name as <path-to-volumes>/<volname>.vmdk"""
+    return os.path.join(path, "%s.vmdk" % vol_name)
```
Comment (andrewjstone) [inline esx_service/utils/vmdk_utils.py]: Nit: I've been trying to stick to the more modern python recommendation of using `"{0}.vmdk".format(vol_name)`, but it doesn't really matter.

File: esx_service/vmdk_ops.py [hunk 3]
```diff
@@ -342,51 +347,109 @@ def detachVMDK(vmdk_path, vm_uuid):
...
+def get_vol_path(datastore):
+    # The folder for Docker volumes is created on <datastore>/DOCK_VOLS_DIR
+    path = os.path.join("/vmfs/volumes", datastore, DOCK_VOLS_DIR)
+    
     if os.path.isdir(path):
         # If the path exists then return it as is.
         logging.debug("Found %s, returning", path)
         return path
-
+
```
Comment (andrewjstone) [inline esx_service/vmdk_ops.py]: Nit: Please don't add extra whitespace chars

Comment (msterin) [inline esx_service/vmdk_ops.py]: lack of these spaces screws up python auto-formatting in editor and external tools....

File: esx_service/vmdk_ops.py [hunk 4]
```diff
@@ -342,51 +347,109 @@ def detachVMDK(vmdk_path, vm_uuid):
...
+        return vol_name
+    return "{0}@{1}".format(vol_name, datastore)
+
+
+# TBD - move to vmdk_utils
+def get_datastore_name(config_path):
+    """Returns datastore NAME in config_path (not url-name which may be used in path)"""
+    # path is always /vmfs/volumes/<datastore>/... , so extract datastore:
+    config_ds_name = config_path.split("/")[3]
+    ds_name = [x[0] for x in vmdk_utils.get_datastores() \
```
Comment (andrewjstone) [inline esx_service/vmdk_ops.py]: When would url-name == config_ds_name ? Why not just check x[0] (name)

Comment (msterin) [inline esx_service/vmdk_ops.py]: it should not, but I felt it's more reliable this way :-)  - no harm in being paranoid

File: esx_service/vmdk_ops.py (hunk 3 above)
Comment (andrewjstone) [inline esx_service/vmdk_ops.py]: Weird, but ok.

**Comment by andrewjstone** [comment]
LGTM

File: esx_service/utils/vmdk_utils.py [hunk 5]
```diff
@@ -50,19 +55,23 @@ def get_datastores():
...
-    for (datastore, path) in get_datastores():
+    for (datastore, url_name, path) in get_datastores():
         for file_name in list_vmdks(path):
             volumes.append({'path': path,
                             'filename': file_name,
                             'datastore': datastore})
     return volumes
 
 
+def get_vmdk_name(path, vol_name):
```
Comment (pdhamdhere) [inline esx_service/utils/vmdk_utils.py]: nit: Should this be really get_vmdk_path?

File: esx_service/vmdk_ops.py [hunk 6]
```diff
@@ -342,51 +347,109 @@ def detachVMDK(vmdk_path, vm_uuid):
...
+def known_datastores():
+    """returns names of know datastores"""
+    return [i[0] for i in vmdk_utils.get_datastores()]
 
+def parse_vol_name(full_vol_name):
+    """
+    Parses volume[@datastore] and returns (volume, datastore)
+    On parse errors raises ValidationError with syntax explanation
+    """
+    groups = re.match("\A([\w\-.]+)@?(\w+)?$", full_vol_name)
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: So, with this change "@" has special meaning and can't be used in volume name. Update FAQ on wiki?

On a side note, docker volume create -d vmdk --name a/b` fails with
```
Error response from daemon: create a/b: VolumeDriver.Create: Failed to create /vmfs/volumes/vsan:521db3284facc204-03f387b708983c04/dockvols/a/b.vmdk. Failed to create virtual disk: 
```

Comment (msterin) [inline esx_service/vmdk_ops.py]: I see this :  
```
root@photon-machine [ ~ ]# docker volume create -d vmdk --name=a/b        
Error response from daemon: create a/b: VolumeDriver.Create: u"Invalid syntax: 'a/b'\nValid syntax is volume@datastore, where volume is [a-zA-Z0-9_-. and datastore is [a-zA-Z0-9_]"
```

File: esx_service/utils/vmdk_utils.py (hunk 5 above)
Comment (msterin) [inline esx_service/utils/vmdk_utils.py]: sure

File: esx_service/vmdk_ops.py [hunk 7]
```diff
@@ -342,51 +347,109 @@ def detachVMDK(vmdk_path, vm_uuid):
...
+    if not groups:
+        raise ValidationError("Invalid syntax: '%s'. " \
+                           "Valid syntax is volume@datastore, where volume is " \
+                           "[a-zA-Z0-9_-. and datastore is [a-zA-Z0-9_]" %
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: Update regular expression in error message to match one on L380

File: esx_service/vmdk_ops.py (hunk 6 above)
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: Yes, with this change. I don't have your change.

Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: I don't know what are naming constraints on datastore name. Checking if you have considered same.

**Comment by pdhamdhere** [comment]
LGTM.

File: esx_service/utils/vmdk_utils.py (hunk 2 above)
Comment (msterin) [inline esx_service/utils/vmdk_utils.py]: fixed

File: esx_service/vmdk_ops.py (hunk 7 above)
Comment (msterin) [inline esx_service/vmdk_ops.py]: made both to [a-zA-Z0-9_-.]+

**Pull request merged**